src/main/scala/firrtl/stage/FirrtlStage.scala

```diff
@@ -0,0 +1,105 @@
+// See LICENSE for license details.
+
+package firrtl.stage
+
+import firrtl._
+
+import firrtl.options.{Stage, Phase, StageUtils, Shell, OptionsException, StageOptions}
+import firrtl.options.Viewer.view
+import firrtl.annotations.{Annotation, NoTargetAnnotation, JsonProtocol, DeletedAnnotation}
+import firrtl.passes.{PassException, PassExceptions}
+import firrtl.transforms.BlackBoxTargetDirAnno
+import firrtl.Utils.throwInternalError
+
+import scala.util.{Failure, Success}
+import scala.util.control.ControlThrowable
+
+import logger.{Logger, LoggerCli}
+
+case class FirrtlStageException(message: String) extends Exception(message)
+
+sealed trait FirrtlStageResult extends Product
+case class FirrtlStageSuccess(emitType: String, emitted: String, circuitState: CircuitState) extends NoTargetAnnotation
+    with FirrtlStageResult
+case class FirrtlStageFailure(message: String) extends NoTargetAnnotation with FirrtlStageResult
+
+object FirrtlStage extends Stage {
+  val shell: Shell = new Shell("firrtl") with FirrtlCli with LoggerCli
+
+  private val phases: Seq[Phase] = Seq(
+    firrtl.stage.phases.AddDefaults,
+    firrtl.stage.phases.Checks,
+    firrtl.stage.phases.AddEmitter,
+    firrtl.stage.phases.AddCircuit,
+    firrtl.stage.phases.AddImplicitOutputFile )
+
+  def run(inputAnnotations: AnnotationSeq): AnnotationSeq = Logger.makeScope(inputAnnotations) {
+    val annotations = try {
+      phases.foldLeft(inputAnnotations)((a, f) => f.transform(a))
+    } catch {
+      case OptionsException(msg, _) =>
+        if (inputAnnotations.collectFirst{ case a: DriverCompatibilityAnnotation.type => a }.isEmpty) {
+          StageUtils.dramaticError(msg)
+          shell.parser.showUsage
+        }
+        return FirrtlStageFailure(msg) +: inputAnnotations
+      case e: Exception => throwInternalError(exception = Some(e))
+    }
+
+    val firrtlOptions = view[FirrtlOptions](annotations)
+    val stageOptions = view[StageOptions](annotations)
+
+    // Wrap compilation in a try/catch to present Scala MatchErrors in a more user-friendly format.
+    val finalState = try {
+      val circuit = firrtlOptions.firrtlCircuit.getOrElse { // This should be impossible
+        throwInternalError("FirrtlStage annotation preprocessing ok, but firrtlOptions had no firrtlCircuit!") }
+
+      firrtlOptions.compiler.compile(
+        CircuitState(circuit, ChirrtlForm, firrtlOptions.annotations),
+        firrtlOptions.customTransforms
+      )
+    }
+    catch {
+      // Rethrow the exceptions which are expected or due to the runtime environment (out of memory, stack overflow)
+      case p @ (_: ControlThrowable | _: PassException | _: PassExceptions | _: FIRRTLException) =>
+        throw p
+      // Treat remaining exceptions as internal errors.
+      case e: Exception => throwInternalError(exception = Some(e))
+    }
+
+    // Do emission
+    val emittedRes = firrtlOptions.getOutputConfig match {
```

Comment:
getOutputConfig should have parens

Reply:
Leaving this as is since I removed the parens from `getOutputConfig` in `FirrtlOptions`.